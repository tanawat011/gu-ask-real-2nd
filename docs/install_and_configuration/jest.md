# Jest for typescript

```text
`Jest` is a test runner.
```

## Documentation

- [Jest.io](https://jestjs.io/docs/getting-started)
- [testing-library](https://testing-library.com/docs/)

---

## Install and Configuration

### Install Jest for typescript

```sh
yarn add jest ts-jest
```

### Configuration

#### Set up test script in your [`package.json`](../package.json) file

```diff
...
 "scripts": {
    ...
+    "test": "react-scripts test --watchAll=false",
+    "test:w": "react-scripts test",
+    "test:cv": "yarn test --coverage",
+    "test:cv:msg": "yarn test:cv --verbose",
+    "test:cv:w": "yarn test:w --coverage",
+    "test:cv:msg:w": "yarn test:cv:w --verbose",
+    "test:cv:open": "open coverage/lcov-report/index.html",
    ...
  },
...
```

#### Set up basic configuration in your [`package.json`](../package.json) file

```diff
...
+  "jest": {
+    "coverageThreshold": {
+      "global": {
+        "statements": 40,
+        "branches": 40,
+        "functions": 40,
+        "lines": 40
+      }
+    },
+    "coveragePathIgnorePatterns": [
+      "src/pages/*",
+      "src/types/*",
+      "index.tsx",
+      "index.ts"
+    ]
+  }
...
```
