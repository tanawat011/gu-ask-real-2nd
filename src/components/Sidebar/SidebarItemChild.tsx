import type { SidebarState } from 'recoils/atoms'

import tw, { styled } from 'twin.macro'

import { DARK_THEME, LIGHT_THEME } from 'constants/twTheme'

export type SidebarItemProps = {
  label: string
  onClick?: () => void
  selected?: boolean
  fullSidebar?: boolean
}

type SidebarItemContainerProps = Pick<SidebarItemProps, 'selected'> & SidebarState

const TwContainer = styled.div(({ fullSidebar, selected }: SidebarItemContainerProps) => [
  !selected && tw`ml-5 pl-3!`,
  fullSidebar
    ? [
        [DARK_THEME.SIDEBAR.MENU.TEXT_COLOR],
        [DARK_THEME.SIDEBAR.MENU.HOVER.BG_COLOR, DARK_THEME.SIDEBAR.MENU.HOVER.TEXT_COLOR],
        [LIGHT_THEME.SIDEBAR.MENU.TEXT_COLOR],
        [LIGHT_THEME.SIDEBAR.MENU.HOVER.BG_COLOR, LIGHT_THEME.SIDEBAR.MENU.HOVER.TEXT_COLOR],
        tw`h-10 pl-3`,
      ]
    : [
        [DARK_THEME.MINI_SIDEBAR.MENU.TEXT_COLOR],
        [
          DARK_THEME.MINI_SIDEBAR.MENU.HOVER.BG_COLOR,
          DARK_THEME.MINI_SIDEBAR.MENU.HOVER.TEXT_COLOR,
        ],
        [LIGHT_THEME.MINI_SIDEBAR.MENU.TEXT_COLOR],
        [
          LIGHT_THEME.MINI_SIDEBAR.MENU.HOVER.BG_COLOR,
          LIGHT_THEME.MINI_SIDEBAR.MENU.HOVER.TEXT_COLOR,
        ],
        tw`ml-0 h-9 px-4 py-1`,
      ],
  tw`flex cursor-pointer items-center justify-between rounded-lg`,
])
const TwItem = tw.div`flex items-center`
const TwItemSelected = styled.div(() => [
  [DARK_THEME.SIDEBAR.MENU.ACTIVE.BG_COLOR, DARK_THEME.SIDEBAR.MENU.ACTIVE.TEXT_COLOR],
  [LIGHT_THEME.SIDEBAR.MENU.ACTIVE.BG_COLOR, LIGHT_THEME.SIDEBAR.MENU.ACTIVE.TEXT_COLOR],
  tw`ml-5 rounded-lg pl-0`,
])

export const SidebarItemChild: React.FC<SidebarItemProps> = ({
  label,
  onClick,
  selected,
  fullSidebar,
}) => {
  const renderItem = (
    <TwContainer onClick={onClick} selected={selected} fullSidebar={fullSidebar || false}>
      <TwItem>
        <span className='mt-0.5'>{label}</span>
      </TwItem>
    </TwContainer>
  )

  if (selected) {
    return <TwItemSelected>{renderItem}</TwItemSelected>
  }

  return renderItem
}
