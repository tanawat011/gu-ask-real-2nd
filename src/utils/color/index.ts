export * from './hex2Rgb'
export * from './rgb2Hex'
