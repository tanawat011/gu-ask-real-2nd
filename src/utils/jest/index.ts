export * from './renderWithProviders'
export * from './twColor'
