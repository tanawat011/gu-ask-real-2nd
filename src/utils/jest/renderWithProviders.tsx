import type { RenderResult } from '@testing-library/react'

import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
import { render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { RecoilRoot } from 'recoil'

const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  const queryClient = new QueryClient()

  return (
    <QueryClientProvider client={queryClient}>
      <RecoilRoot>
        <BrowserRouter>{children}</BrowserRouter>
      </RecoilRoot>
    </QueryClientProvider>
  )
}

interface RenderWithProvidersOptions {
  router?: {
    route: string
  }
  recoil?: object
}

export const renderWithProviders = (
  ui: JSX.Element,
  option?: RenderWithProvidersOptions,
): RenderResult => {
  const routerOption = { ...(option?.router || { route: '/' }) }
  window.history.pushState({}, 'Test page', routerOption.route)

  return {
    ...render(ui, { wrapper: AllTheProviders, ...routerOption }),
  }
}
