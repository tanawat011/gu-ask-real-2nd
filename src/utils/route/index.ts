export * from './menuGenerator'
export * from './renderRoutes'
export * from './routeGenerator'
