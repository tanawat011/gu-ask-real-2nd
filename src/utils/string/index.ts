export * from './camelToKebab'
export * from './camelToSentence'
export * from './replaceAll'
