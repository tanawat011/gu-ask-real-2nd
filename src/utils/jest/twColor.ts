import { hex2Rgb } from 'utils/color'

export const twColor = (hexColor: string, type?: 'bg' | 'border' | 'text') => {
  if (hexColor === 'transparent') {
    return 'transparent'
  }

  if (hexColor.length > 7) {
    return 'undefined'
  }

  const rgb = hex2Rgb(hexColor, { isArray: true }) as string[]

  if (!type) {
    type = 'bg'
  }

  return `rgb(${rgb.join(' ')} / var(--tw-${type}-opacity))`
}
