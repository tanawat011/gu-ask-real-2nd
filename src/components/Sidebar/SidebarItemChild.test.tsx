import { renderWithProviders } from 'utils/jest'

import { SidebarItemChild } from './SidebarItemChild'

jest.mock('./SidebarItemChild', () => jest.requireActual('./SidebarItemChild'))

describe('<SidebarItemChild />', () => {
  test('should render without crashing and match snapshot', () => {
    expect.hasAssertions()

    const { asFragment } = renderWithProviders(
      <SidebarItemChild label='test' onClick={jest.fn()} fullSidebar />,
    )

    expect(asFragment()).toMatchSnapshot()
  })

  test('should render without crashing and mini sidebar', () => {
    expect.hasAssertions()

    const { container } = renderWithProviders(
      <SidebarItemChild label='test' onClick={jest.fn()} fullSidebar={false} />,
    )

    expect(container).toBeInTheDocument()
  })

  test('should render without crashing and selected item', () => {
    expect.hasAssertions()

    const { container } = renderWithProviders(
      <SidebarItemChild
        label='test'
        onClick={jest.fn()}
        selected
        fullSidebar
      />,
    )

    expect(container).toBeInTheDocument()
  })
})
