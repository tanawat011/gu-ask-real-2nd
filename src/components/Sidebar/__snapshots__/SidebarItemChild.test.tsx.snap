// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SidebarItemChild /> should render without crashing and match snapshot 1`] = `
<DocumentFragment>
  .c1 {
  margin-left: 1.25rem;
  padding-left: 0.75rem !important;
  --tw-text-opacity: 1;
  color: rgb(75 85 99 / var(--tw-text-opacity));
  height: 2.5rem;
  padding-left: 0.75rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border-radius: 0.5rem;
}

.dark .c0 {
  --tw-text-opacity: 1;
  color: rgb(156 163 175 / var(--tw-text-opacity));
}

.dark .c0:hover {
  --tw-bg-opacity: 1;
  background-color: rgb(41 51 65 / var(--tw-bg-opacity));
}

.dark .c0:hover {
  --tw-text-opacity: 1;
  color: rgb(255 255 255 / var(--tw-text-opacity));
}

.c1:hover {
  --tw-bg-opacity: 0.1;
  background-color: rgb(17 24 39 / var(--tw-bg-opacity));
}

.c1:hover {
  --tw-text-opacity: 1;
  color: rgb(17 24 39 / var(--tw-text-opacity));
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
    class="c0 c1"
  >
    <div
      class="c2"
    >
      <span
        class="mt-0.5"
      >
        test
      </span>
    </div>
  </div>
</DocumentFragment>
`;
